### ISSUE TITLE
for of does not work on generator function when running on es5 aka internet explorer

### ISSUE THREAD

**Issue by Saulzi** [microsoft/TypeScript#32458]
**TypeScript Version:**  3.5.3** 

** Related issue **
https://github.com/microsoft/tslib/issues/59

downleveliteration es5, downleveliteration internet explorer, generator internet explorer

**Code**

```ts
function* issue()
{
    yield "hello";
    yield "world";
}

for (let x of issue())
{
    console.log(x);
}

**Expected behavior:**
hello, world is logged to the console when running on internet explorer 11

**Actual behavior:**
console logs undefined forever as stuck in loop

** Fix **
The code in tslib / which is added to top of transpiled file should check if we are enumerating over an array or have something which is IterableIterator like

``` js
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (m) return m.call(o);
    if (Array.isArray(o)) {
        var i = 0;
        return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
    } else if (typeof o.next === "function") {  // should probably be more concise
        return o;
    }  else {
    // kittens die
    }
};

** Relatd
@microsoft/tslib#59

**Comment by fatcerberus** [comment]
Works for me 🤷‍♂ 

![image](https://user-images.githubusercontent.com/4097841/61414372-c50fab80-a8bb-11e9-99eb-2d852f17f09b.png)

[Playground Link](https://www.typescriptlang.org/play/?downlevelIteration=true&target=1#code/PTAEBcAsEsGdQMYEMB2oBGBTUtoFsAHAG2gDNpMATAKFIFcUFxoB7FAKlDljswAoAlNQDe1UONABPCkUqgARJExEiLeQG4xE6crnyA7iwBOsjdQC+1WsdB8imcKAAeoFqS6we-AUNETEbLAs9gB0qgDmfE4CmpZAA)

**Comment by Saulzi** [comment]
I think I have figured where this is code is defined for injection in the typescript source. 

src/compiler/factory.ts:3635

I am going to attempt to fix this myself and produce a PR which hopefully makes it into TypeScript / tslib.

my thoughts thus far.

__values should do the following when Symbol.iterator is missing i.e. on internet explorer

if the object passed in is an array then use the existing for i loop wrapping in iterable next / done
if the object looks like it is a iterable object, i.e. it has a next and done property then use that otherwise throw a typeerror with the same message as chrome, i.e. x is not iterable, interestingly firefox throws typeerror with undefined has no properties when trying to do for (let x of undefined) {}

``` ts
    export const valuesHelper: UnscopedEmitHelper = {
        name: "typescript:values",
        scoped: false,
        text: `
            var __values = (this && this.__values) || function (o) {
                var m = typeof Symbol === "function" && o[Symbol.iterator];
                if (m) return m.call(o);
                if (typeof o === "object") {
                    if (Array.isArray(o)) {
                        return {
                            next: function () {
                                if (o && i >= o.length) o = void 0;
                                return { value: o && o[i++], done: !o };
                            }
                        };
                    } else if (typeof o.next === "fuction" && typeof o.done === "boolean") {
                        return o;
                    } 
                }
                throw new TypeError(typeof o + " is not iterable");
            };`
    };
```

I am not sure if tslib is automatically generated allthough, if it isn't adding the something like the above to es6 and umd javascript files is trivial.

I will need to update the conformance tests by running gulp baseline-accept.

there needs to be some kind of test for the fallback when symbol.iterator does not exist. at the moment I am not sure how I will be able to test this branch as chances are whatever test is running under will have this defined.

**Comment by AviVahl** [comment]
https://github.com/microsoft/tslib

**Comment by Saulzi** [comment]
Current thoughts / findings are as follows.

__generator return value does not have done property so at the moment I can only test to see if it quacks by looking at o.next.

I think that I should be able to do a stash global symbol object and replace to test, although I am currently not sure how / where I should place the test (under testrunner/factory or similar). Likewise I have not figured out how I can test on emitted code at the moment.

all of the existing does the file look correct tests have been updated, 

I have this working in internet explorer with both customised tslib pulled in via amd (requirejs) and with helpers emitted.

**Comment by rbuckton** [comment]
The semantics here are incorrect. `for..of` and other constructs do not work over Iterators, they only works over Iterables (something having a `Symbol.iterator`). We have an affordance for `Array` and `string` down-level, but shouldn't extend this to just any object with a `next` method. If anything, this should just report an error if `length` is not number. This kind of change to semantics is incompatible with the actual ES2015 semantics and could cause problems for anyone migrating a codebase that might use custom Iterators with this change.

Using `downlevelIteration` requires a polyfill for `Symbol.iterator` (as well as `Set.prototype[Symbol.iterator]`, etc.) when iterating over anything other than arrays or strings. This is likely being masked if you are explicitly adding `es2015.iterable` to your libs in tsconfig.json.

**Comment by Saulzi** [comment]
following on from comments from @rbuckton on https://github.com/microsoft/tslib/pull/69#pullrequestreview-265086770.

I feel that the typescript documentation should be improved to note that `Using downlevelIteration requires a polyfill for Symbol.iterator (as well as Set.prototype[Symbol.iterator], etc.) when iterating over anything other than arrays or strings. i.e. generator functions` 

Likewise this fact should be emphasised in the release notes for the next version of typescript as it has the potential to trip other developers up as support for types from generators has been improved and may mean developers who hadn't used this functionality before will get tripped up the same way as myself with this.

moving on and from @rbuckton 's comments get __values function should effectively do something like the below 
``` ts
            if (o && typeof o.length !== "number") {
                var i = 0;
                return {
                    next: function () {
                        if (o && i >= o.length) o = void 0;
                        return { value: o && o[i++], done: !o };
                    }
                };
            } else if (typeof o === "object" && o.next === "function") {
                throw new TypeError(typeof o + " is iterable but there is no implementation of Symbol / Symbol.iterator (point to some documentation to improve dev experiance?) ");
            }

            throw new TypeError(typeof o + " is not iterable");
```

**Comment by rbuckton** [comment]
> `throw new TypeEror(typeof o + " is iterable but there is no impelementation...`

This is incorrect as an `Iterator` is not inherently iterable (sometimes the case for custom iterators). I don't think the extra message is necessary.

Here's my proposal:

```js
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object not iterable." : "Symbol.iterator is not defined.");
};
```
It's only one line longer than the current helper and gets the message across.

**Comment by Saulzi** [comment]
@rbuckton I think that I figured out the whole unit tests thing for this and have knocked a few together for for-of on es5 which exposed issue with with o.length on 4th line of your proposal, any feedback would be amazing.

**Comment by rbuckton** [comment]
Closed via #32468
